sdk/python/src/dagger/connectors/docker.py

```diff
@@ -66,92 +59,66 @@ def start(self) -> None:
-                " is already in use by container"
-            ) not in e.stdout:
-                raise ProvisionError(f"Failed to start engine container: {e.stdout}")
-
-        # TODO: garbage collection of old containers
-
-        os_, arch = get_platform()
-
-        if not helper_bin_path.exists():
-            tmp_bin = tempfile.NamedTemporaryFile(
-                prefix=f"temp-{HELPER_BINARY_PREFIX}",
-                dir=cache_dir,
-                delete=False,
-            )
-            docker_cp_args = [
-                "docker",
-                "cp",
-                f"{container_name}:/usr/bin/{HELPER_BINARY_PREFIX}{os_}-{arch}",
-                tmp_bin.name,
-            ]
-            try:
-                subprocess.run(
-                    docker_cp_args,
-                    stdout=subprocess.PIPE,
-                    stderr=subprocess.STDOUT,
-                    encoding="utf-8",
-                    check=True,
-                )
-            except subprocess.CalledProcessError as e:
-                tmp_bin.close()
-                os.unlink(tmp_bin.name)
-                raise ProvisionError(f"Failed to copy helper binary: {e.stdout}")
-
-            tmp_bin_path = Path(tmp_bin.name)
-            tmp_bin_path.chmod(0o700)
-
-            helper_bin_path = (
-                tmp_bin_path.rename(helper_bin_path)
-                if image.is_pinned
-                else tmp_bin_path
-            )
-
-        # garbage collection of old helper binaries
-        for bin in cache_dir.glob(f"{HELPER_BINARY_PREFIX}*"):
-            if bin != helper_bin_path:
-                bin.unlink()
-
-        buildkit_host = f"docker-container://{container_name}"
+        engine_session_bin_path = (
+            cache_dir / f"{ENGINE_SESSION_BINARY_PREFIX}{image.id}"
+        )
 
-        helper_args = [helper_bin_path, "--remote", buildkit_host]
+        if not engine_session_bin_path.exists():
+            os_, arch = get_platform()
+            tempfile_args = {
+                "prefix": f"temp-{ENGINE_SESSION_BINARY_PREFIX}",
+                "dir": cache_dir,
+                "delete": False,
+            }
+            with tempfile.NamedTemporaryFile(**tempfile_args) as tmp_bin:
+                docker_run_args = [
+                    "docker",
+                    "run",
+                    "--rm",
+                    "--entrypoint",
+                    "/bin/cat",
+                    image.ref,
+                    f"/usr/bin/{ENGINE_SESSION_BINARY_PREFIX}{os_}-{arch}",
+                ]
+                try:
+                    subprocess.run(
+                        docker_run_args,
+                        stdout=tmp_bin,
+                        stderr=subprocess.PIPE,
+                        encoding="utf-8",
+                        check=True,
+                    )
+                except subprocess.CalledProcessError as e:
+                    tmp_bin.close()
```

Comment:
I see, I used the context manager because I thought that it would handle deleting the file in more cases, but in retrospect I'm not sure if `delete=False` means that delete is also now skipped by the context manager or if it's just skipped when the object is closed (tempfile docs are ambiguous...).

Do you think it's worth skipping the context manager?

Reply:
(I'm gonna merge as is so we can continue with getting the other stuff ready w/ minimal potential for rebase conflicts, but if this is worth a follow up I'll send out more PRs)

Reply:
`delete=False` means it won't attempt to delete the file in any case. Just close.